### PR TITLE
Vote for ideas

### PR DESCRIPTION
**Pull request by mrkvon**
- [x] add a vote to idea
- [x] remove a vote from idea
- [x] show how many up and down votes idea has, show if and how the current user voted
- ~maybe change value of a vote~ (won't do now) (#52)

### CODE REVIEW

File: models/vote/index.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+          FOR v IN votes FILTER v._from == u._id && v._to == i._id
+            RETURN v`;
+    const params = { username, '@type': type, id };
+    const cursor = await this.db.query(query, params);
+
+    return (await cursor.all())[0];
+  }
+
+  /**
+   * Read a vote from user to multiple ideas or comments or something
```
Comment (agatatalita) [inline models/vote/index.js]: I am curious when is it used, the function.

Comment (mrkvon) [reply models/vote/index.js]: This is used to see, whether user voted for a particular comment (to show the self vote). When there are i.e. 20 comments, we want to fetch all the info at once.

File: controllers/votes.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+  try {
+    // save the vote to database
+    const vote = await models.vote.create({ from: username, to: { type: primarys, id }, value });
+    // respond
+    const serializedVote = serializers.serialize.vote(vote);
+    return res.status(201).json(serializedVote);
+  } catch (e) {
+    // handle errors
+    switch (e.code) {
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply controllers/votes.js]: To some extent the error codes are self-explanatory. You're right that it is boring to write (and test) and that some error info would actually be useful. I'll fix it.

File: controllers/votes.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ */
+async function post(req, res, next) {
+
+  // read data from request
+  const { id } = req.params;
+  const { value } = req.body;
+  const { username } = req.auth;
+
+  // what is the type of the object we vote for (i.e. ideas, comments, ...)
+  const primarys = req.baseUrl.substring(1);
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply controllers/votes.js]: No. Do you have some alternative suggestions?
`primary` is supposed to mean "object to which the vote belongs" and `s` is plural

[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/votes.js]: For me primarys can stay it doesn't make a confusion (just lingual one) and it is quite clear. I was mostly curious.

File: controllers/votes.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (agatatalita) [reply controllers/votes.js]: I was thinking more generally if there is any standard in ditup regarding this and what is the explanation.
 Yes, it is quite clear what I am getting if there is 409 or 201, maybe some 400 validation explanation is useful (and that is double boring).

File: models/vote/index.js (hunk 1 above)
Comment (agatatalita) [reply models/vote/index.js]: Thank you.

**Comment by mrkvon** [comment]
Thank you @agatatalita for collaborating on this. :slightly_smiling_face:

**Pull request merged**